File: backend/api/views/employees.py
import json
from django.db.models import Q
from rest_framework import status
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response

from api.models import Employee
from api.permissions import IsAdmin, IsAdminOrEncoder
from api.serializers import EmployeeSerializer, TrainingSerializer
from api.utils import log_audit, today


def _build_full_name(last_name, first_name, middle_initial):
    name = f"{last_name}, {first_name}"
    if middle_initial:
        mi = middle_initial.rstrip('.')
        name += f" {mi}."
    return name


@api_view(['GET'])
@permission_classes([IsAuthenticated])
def employee_filters(request):
    factories = list(
        Employee.objects.exclude(factory='')
        .values_list('factory', flat=True)
        .distinct()
        .order_by('factory')
    )
    lines = list(
        Employee.objects.exclude(line='')
        .values_list('line', flat=True)
        .distinct()
        .order_by('line')
    )
    teams = list(
        Employee.objects.exclude(team='')
        .values_list('team', flat=True)
        .distinct()
        .order_by('team')
    )
    return Response({'factories': factories, 'lines': lines, 'teams': teams})


@api_view(['GET', 'POST'])
@permission_classes([IsAuthenticated])
def employee_list_create(request):
    if request.method == 'GET':
        qs = Employee.objects.all()
        for field in ('status', 'factory', 'line', 'team', 'employment_status'):
            value = request.query_params.get(field)
            if value:
                qs = qs.filter(**{field: value})
        search = request.query_params.get('search')
        if search:
            qs = qs.filter(
                Q(full_name__icontains=search) | Q(employee_id__icontains=search)
            )
        return Response(EmployeeSerializer(qs.order_by('full_name'), many=True).data)

    if not IsAdminOrEncoder().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=status.HTTP_403_FORBIDDEN)

    employee_id = (request.data.get('employee_id') or '').strip()
    last_name = (request.data.get('last_name') or '').strip()
    first_name = (request.data.get('first_name') or '').strip()
    middle_initial = (request.data.get('middle_initial') or '').strip()

    if not employee_id or not last_name or not first_name:
        return Response(
            {'error': 'employee_id, last_name, and first_name are required'},
            status=status.HTTP_400_BAD_REQUEST,
        )

    if Employee.objects.filter(employee_id=employee_id).exists():
        return Response({'error': 'Employee ID already exists'}, status=status.HTTP_409_CONFLICT)

    full_name = _build_full_name(last_name, first_name, middle_initial)

    employee = Employee.objects.create(
        employee_id=employee_id,
        last_name=last_name,
        first_name=first_name,
        middle_initial=middle_initial,
        full_name=full_name,
        factory=request.data.get('factory') or '',
        line=request.data.get('line') or '',
        team=request.data.get('team') or '',
        position=request.data.get('position') or '',
        employment_status=request.data.get('employment_status') or '',
        status=request.data.get('status') or 'active',
        hire_date=request.data.get('hire_date') or None,
    )
    log_audit(request.user, 'CREATE', 'employees', employee.id, f'Created employee: {full_name}')
    return Response(EmployeeSerializer(employee).data, status=status.HTTP_201_CREATED)


@api_view(['GET', 'PUT', 'DELETE'])
@permission_classes([IsAuthenticated])
def employee_detail(request, pk):
    try:
        employee = Employee.objects.get(pk=pk)
    except Employee.DoesNotExist:
        return Response({'error': 'Employee not found'}, status=status.HTTP_404_NOT_FOUND)

    if request.method == 'GET':
        data = EmployeeSerializer(employee).data
        trainings = employee.trainings.order_by('-training_date')
        data['trainings'] = TrainingSerializer(trainings, many=True).data
        return Response(data)

    if request.method == 'PUT':
        if not IsAdminOrEncoder().has_permission(request, None):
            return Response({'error': 'Insufficient permissions'}, status=status.HTTP_403_FORBIDDEN)

        # Capture before state
        before = {
            'employee_id': employee.employee_id,
            'full_name': employee.full_name,
            'factory': employee.factory,
            'line': employee.line,
            'team': employee.team,
            'position': employee.position,
            'employment_status': employee.employment_status,
            'hire_date': str(employee.hire_date) if employee.hire_date else None,
            'status': employee.status,
        }

        employee_id = (request.data.get('employee_id') or employee.employee_id).strip()
        if not employee_id:
            return Response(
                {'error': 'employee_id is required'},
                status=status.HTTP_400_BAD_REQUEST,
            )
        if employee_id != employee.employee_id:
            if Employee.objects.filter(employee_id=employee_id).exclude(pk=employee.pk).exists():
                return Response({'error': 'Employee ID already exists'}, status=status.HTTP_409_CONFLICT)
            employee.employee_id = employee_id

        last_name = (request.data.get('last_name') or employee.last_name).strip()
        first_name = (request.data.get('first_name') or employee.first_name).strip()
        middle_initial = request.data.get('middle_initial', employee.middle_initial).strip()

        employee.last_name = last_name
        employee.first_name = first_name
        employee.middle_initial = middle_initial
        employee.full_name = _build_full_name(last_name, first_name, middle_initial)

        for field in ('factory', 'line', 'team', 'position', 'employment_status', 'hire_date'):
            if field in request.data:
                setattr(employee, field, request.data[field])
        if request.data.get('status'):
            employee.status = request.data['status']
        employee.save()

        after = {
            'employee_id': employee.employee_id,
            'full_name': employee.full_name,
            'factory': employee.factory,
            'line': employee.line,
            'team': employee.team,
            'position': employee.position,
            'employment_status': employee.employment_status,
            'hire_date': str(employee.hire_date) if employee.hire_date else None,
            'status': employee.status,
        }
        changes = {k: {'before': before[k], 'after': after[k]} for k in before if before[k] != after[k]}

        log_audit(
            request.user, 'UPDATE', 'employees', employee.id,
            json.dumps({'summary': f'Updated employee: {employee.full_name}', 'changes': changes}),
        )
        return Response(EmployeeSerializer(employee).data)

    if not IsAdmin().has_permission(request, None):
        return Response({'error': 'Insufficient permissions'}, status=status.HTTP_403_FORBIDDEN)

    employee.status = 'resigned'
    employee.save()
    log_audit(
        request.user, 'DELETE', 'employees', employee.id,
        f'Deactivated employee: {employee.full_name}',
    )
    return Response({'message': 'Employee deactivated'})

File: client/electron/db/employees.js
'use strict';

const { getPool } = require('./pool');
const { logAudit } = require('./audit');
const { requireAuth, requireRole } = require('./auth');
const { apiError, dt } = require('./helpers');

function serialize(row) {
  return {
    id: row.id,
    employee_id: row.employee_id,
    last_name: row.last_name,
    first_name: row.first_name,
    middle_initial: row.middle_initial,
    full_name: row.full_name,
    factory: row.factory,
    line: row.line,
    team: row.team,
    employment_status: row.employment_status,
    status: row.status,
    hire_date: row.hire_date || null,
    created_at: dt(row.created_at),
    updated_at: dt(row.updated_at),
  };
}

function buildFullName(lastName, firstName, middleInitial) {
  let name = `${lastName}, ${firstName}`;
  if (middleInitial) {
    const mi = middleInitial.replace(/\.+$/, '');
    name += ` ${mi}.`;
  }
  return name;
}

async function getById(id) {
  const pool = getPool();
  const [rows] = await pool.query('SELECT * FROM employees WHERE id = ? LIMIT 1', [id]);
  return rows[0] || null;
}

async function list(params = {}) {
  requireAuth();
  const pool = getPool();
  const where = [];
  const vals = [];
  for (const field of ['status', 'factory', 'line', 'team', 'employment_status']) {
    if (params[field]) {
      where.push(`${field} = ?`);
      vals.push(params[field]);
    }
  }
  if (params.search) {
    where.push('(full_name LIKE ? OR employee_id LIKE ?)');
    vals.push(`%${params.search}%`, `%${params.search}%`);
  }
  let sql = 'SELECT * FROM employees';
  if (where.length) sql += ' WHERE ' + where.join(' AND ');
  sql += ' ORDER BY full_name';
  const [rows] = await pool.query(sql, vals);
  return rows.map(serialize);
}

async function filters() {
  requireAuth();
  const pool = getPool();
  const [f] = await pool.query("SELECT DISTINCT factory FROM employees WHERE factory <> '' ORDER BY factory");
  const [l] = await pool.query("SELECT DISTINCT line FROM employees WHERE line <> '' ORDER BY line");
  const [t] = await pool.query("SELECT DISTINCT team FROM employees WHERE team <> '' ORDER BY team");
  return {
    factories: f.map((r) => r.factory),
    lines: l.map((r) => r.line),
    teams: t.map((r) => r.team),
  };
}

async function get({ id } = {}) {
  requireAuth();
  const pool = getPool();
  const row = await getById(id);
  if (!row) throw apiError('Employee not found', 404);
  const data = serialize(row);
  const trainings = require('./trainings');
  data.trainings = await trainings.listForEmployee(row.id);
  return data;
}

async function create(data = {}) {
  const user = requireRole('admin', 'encoder');
  const pool = getPool();

  const employeeId = (data.employee_id || '').trim();
  const lastName = (data.last_name || '').trim();
  const firstName = (data.first_name || '').trim();
  const middleInitial = (data.middle_initial || '').trim();

  if (!employeeId || !lastName || !firstName) {
    throw apiError('employee_id, last_name, and first_name are required', 400);
  }

  const [exists] = await pool.query('SELECT id FROM employees WHERE employee_id = ? LIMIT 1', [employeeId]);
  if (exists.length) throw apiError('Employee ID already exists', 409);

  const fullName = buildFullName(lastName, firstName, middleInitial);
  const [res] = await pool.query(
    `INSERT INTO employees
      (employee_id, last_name, first_name, middle_initial, full_name, factory, line, team,
       employment_status, status, hire_date, created_at, updated_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NOW(6), NOW(6))`,
    [
      employeeId, lastName, firstName, middleInitial, fullName,
      data.factory || '', data.line || '', data.team || '',
      data.employment_status || '',
      data.status || 'active', data.hire_date || null,
    ]
  );
  const id = res.insertId;
  await logAudit(user, 'CREATE', 'employees', id, `Created employee: ${fullName}`);
  const row = await getById(id);
  return serialize(row);
}

async function update({ id, data } = {}) {
  const user = requireRole('admin', 'encoder');
  const pool = getPool();
  const employee = await getById(id);
  if (!employee) throw apiError('Employee not found', 404);

  const snapshot = (e) => ({
    employee_id: e.employee_id,
    full_name: e.full_name,
    factory: e.factory,
    line: e.line,
    team: e.team,
    employment_status: e.employment_status,
    hire_date: e.hire_date || null,
    status: e.status,
  });
  const before = snapshot(employee);
  data = data || {};

  let employeeId = (data.employee_id || employee.employee_id).trim();
  if (!employeeId) throw apiError('employee_id is required', 400);
  if (employeeId !== employee.employee_id) {
    const [dup] = await pool.query('SELECT id FROM employees WHERE employee_id = ? AND id <> ? LIMIT 1', [employeeId, id]);
    if (dup.length) throw apiError('Employee ID already exists', 409);
  }

  const lastName = (data.last_name || employee.last_name || '').trim();
  const firstName = (data.first_name || employee.first_name || '').trim();
  let middleInitial = ('middle_initial' in data ? String(data.middle_initial ?? '') : (employee.middle_initial || '')).trim();
  const fullName = buildFullName(lastName, firstName, middleInitial);

  const next = {
    employee_id: employeeId,
    last_name: lastName,
    first_name: firstName,
    middle_initial: middleInitial,
    full_name: fullName,
    factory: employee.factory,
    line: employee.line,
    team: employee.team,
    employment_status: employee.employment_status,
    hire_date: employee.hire_date || null,
    status: employee.status,
  };

  for (const field of ['factory', 'line', 'team', 'employment_status', 'hire_date']) {
    if (field in data) {
      let value = data[field];
      if (field === 'hire_date' && (value === '' || value === undefined)) value = null;
      next[field] = value;
    }
  }
  if (data.status) next.status = data.status;

  await pool.query(
    `UPDATE employees SET
       employee_id = ?, last_name = ?, first_name = ?, middle_initial = ?, full_name = ?,
       factory = ?, line = ?, team = ?, employment_status = ?, status = ?,
       hire_date = ?, updated_at = NOW(6)
     WHERE id = ?`,
    [
      next.employee_id, next.last_name, next.first_name, next.middle_initial, next.full_name,
      next.factory, next.line, next.team, next.employment_status, next.status,
      next.hire_date, id,
    ]
  );

  const after = snapshot({ ...next });
  const changes = {};
  for (const key of Object.keys(before)) {
    if (before[key] !== after[key]) changes[key] = { before: before[key], after: after[key] };
  }
  await logAudit(
    user, 'UPDATE', 'employees', id,
    JSON.stringify({ summary: `Updated employee: ${fullName}`, changes })
  );

  const row = await getById(id);
  return serialize(row);
}

async function remove({ id } = {}) {
  const user = requireRole('admin');
  const pool = getPool();
  const employee = await getById(id);
  if (!employee) throw apiError('Employee not found', 404);
  await pool.query("UPDATE employees SET status = 'resigned', updated_at = NOW(6) WHERE id = ?", [id]);
  await logAudit(user, 'DELETE', 'employees', id, `Deactivated employee: ${employee.full_name}`);
  return { message: 'Employee deactivated' };
}

module.exports = { list, filters, get, create, update, remove, serialize, getById };
